### PR TITLE
Lotus 58, Vial keymap

### PR DESCRIPTION
**Pull request by TweetyDaBird**
Adding Vial keymap, updated to latest QMK keycodes and moved appropriate settings to info.JSON

BTW, the two options failing in compile is due to lack of space using default bootloaders, and aren't viable as targets for Vial. So either these can be ignored, or someone can show me how to 'turn them off'?

### CODE REVIEW

File: keyboards/tweetydabird/lotus58/keymaps/vial/keymap.c [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    ),
+
+    [2] = LAYOUT(
+    ____, ____, ____, ____, ____, ____, ____,        ____, ____, ____, ____, ____, ____, ____,
+	____, ____, ____, ____, ____, ____,                    ____, ____, ____, ____, ____, ____,
+	____, ____, ____, ____, ____, ____,                    ____, ____, ____, ____, ____, ____,
+	____, ____, ____, ____, ____, ____, ____,        ____, ____, ____, ____, ____, ____, ____,
+					  ____, ____, ____, ____,      	 ____, ____, ____, ____
+    ),
+
```
Comment (lesshonor) [inline keyboards/tweetydabird/lotus58/keymaps/vial/keymap.c]: ...Is wedging in this blank layer really necessary?

File: keyboards/tweetydabird/lotus58/keymaps/vial/keymap.c [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+					  ____, ____, ____, ____,      	 ____, ____, ____, ____
+    ),
+};
+
+#if defined(ENCODER_MAP_ENABLE)
+const uint16_t PROGMEM encoder_map[][NUM_ENCODERS][2] = {
+    [0] = { ENCODER_CCW_CW(KC_VOLD, KC_VOLU), ENCODER_CCW_CW(KC_VOLD, KC_VOLU) },
+    [1] = { ENCODER_CCW_CW(KC_VOLD, KC_VOLU), ENCODER_CCW_CW(KC_VOLD, KC_VOLU) },
+    [2] = { ENCODER_CCW_CW(KC_VOLD, KC_VOLU), ENCODER_CCW_CW(KC_VOLD, KC_VOLU) },
+    [3] = { ENCODER_CCW_CW(KC_VOLD, KC_VOLU), ENCODER_CCW_CW(KC_VOLD, KC_VOLU) }
```
Comment (lesshonor) [inline keyboards/tweetydabird/lotus58/keymaps/vial/keymap.c]:
```suggestion
    [1] = { ENCODER_CCW_CW(KC_TRNS, KC_TRNS), ENCODER_CCW_CW(KC_TRNS, KC_TRNS) },
    [2] = { ENCODER_CCW_CW(KC_TRNS, KC_TRNS), ENCODER_CCW_CW(KC_TRNS, KC_TRNS) },
    [3] = { ENCODER_CCW_CW(KC_TRNS, KC_TRNS), ENCODER_CCW_CW(KC_TRNS, KC_TRNS) }
```

File: keyboards/tweetydabird/lotus58/keymaps/vial/keymap.c [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+#endif
+
+#ifdef OLED_ENABLE
+
+static void print_status_narrow(void) {
+	// Create OLED content
+    oled_write_P(PSTR("\n"), false);
+	oled_write_P(PSTR(""), false);
+    oled_write_P(PSTR("Lotus -58-"), false);
+	oled_write_P(PSTR("\n"), false);
```
Comment (lesshonor) [inline keyboards/tweetydabird/lotus58/keymaps/vial/keymap.c]:
```suggestion
    oled_advance_page(true);
    oled_write_P(PSTR("Lotus -58-"), false);
    oled_advance_page(true);
```
Not sure exactly what the point of the blank string is, but using `advance_page` here will improve matrix scan rate (and means you don't have to store additional line breaks in `PROGMEM`)

If you know there will never be text in any of these areas (so they will never have to be cleared) you could use `oled_advance_page(false)` to eke out another tiny efficiency boost.

File: keyboards/tweetydabird/lotus58/keymaps/vial/keymap.c [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+            break;
+        case 3:
+            oled_write_P(PSTR("-Sys \n"), false);
+            break;
+        default:
+            oled_write_P(PSTR("Undef"), false);
+    }
+	
+	// Display capslock
+    oled_write_P(PSTR("\n"), false);
```
Comment (lesshonor) [inline keyboards/tweetydabird/lotus58/keymaps/vial/keymap.c]:
```suggestion
    oled_advance_page(true);
```

File: keyboards/tweetydabird/lotus58/keymaps/vial/keymap.c [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    led_t led_usb_state = host_keyboard_led_state();
+    oled_write_ln_P(PSTR("Caps- lock"), led_usb_state.caps_lock);
+	
+#ifdef AUTO_SHIFT_ENABLE
+
+	bool autoshift = get_autoshift_state();
+	oled_write_P(PSTR("\n"), false);
+	oled_write_P(PSTR("Auto-Shift"), autoshift);
+	oled_write_P(PSTR("\n"), false);
```
Comment (lesshonor) [inline keyboards/tweetydabird/lotus58/keymaps/vial/keymap.c]:
```suggestion
	oled_advance_page(true);
	oled_write_P(PSTR("Auto-Shift"), autoshift);
	oled_advance_page(true);
```

File: keyboards/tweetydabird/lotus58/keymaps/vial/keymap.c [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+}
+
+oled_rotation_t oled_init_user(oled_rotation_t rotation) {
+    if (is_keyboard_left()) {
+        return OLED_ROTATION_270;
+    } else {
+		return OLED_ROTATION_270;
+	}
+    return rotation;
```
Comment (lesshonor) [inline keyboards/tweetydabird/lotus58/keymaps/vial/keymap.c]:
```suggestion
	return OLED_ROTATION_270;
```

File: keyboards/tweetydabird/lotus58/keymaps/vial/keymap.c [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+bool oled_task_user(void) {
+    if (is_keyboard_master()) {
+        print_status_narrow();
+    } else {
+        print_status_narrow();
```
Comment (lesshonor) [inline keyboards/tweetydabird/lotus58/keymaps/vial/keymap.c]:
```suggestion
```

File: keyboards/tweetydabird/lotus58/keymaps/vial/keymap.c [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    }
+	return false;
+}
+
+#endif
+
+layer_state_t layer_state_set_user(layer_state_t state) {
+   return update_tri_layer_state(state, 1, 2, 3);
+}
```
Comment (lesshonor) [inline keyboards/tweetydabird/lotus58/keymaps/vial/keymap.c]:
```suggestion
```
Locking out layers with no way to determine what's going on from the Vial GUI definitely results in confusion. Users are being primed to look for [the tri-layer paradigm laid out in the QMK docs](https://docs.qmk.fm/#/feature_tri_layer) versus a `layer_state` callback.

**Review by lesshonor — CHANGES_REQUESTED** [review]

File: keyboards/tweetydabird/lotus58/keymaps/vial/keymap.c (hunk 1 above)
Comment (TweetyDaBird) [reply keyboards/tweetydabird/lotus58/keymaps/vial/keymap.c]: Nope, just a leftover from Lily58 firmware which I started with years ago.

File: keyboards/tweetydabird/lotus58/keymaps/vial/keymap.c (hunk 8 above)
Comment (TweetyDaBird) [reply keyboards/tweetydabird/lotus58/keymaps/vial/keymap.c]: Agreed, that was a carryover from QMK, so it's going away.

File: keyboards/tweetydabird/lotus58/keymaps/vial/keymap.c [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+oled_rotation_t oled_init_user(oled_rotation_t rotation) {
+	return OLED_ROTATION_270;
+}
+
+bool oled_task_user(void) {
+    if (is_keyboard_master()) {
+        print_status_narrow();
+    }
```
Comment (lesshonor) [inline keyboards/tweetydabird/lotus58/keymaps/vial/keymap.c]: Terribly sorry, I steered you slightly wrong here with my earlier suggestion. As you're printing the same thing on both sides, the entire `if` statement isn't necessary.
```suggestion
    print_status_narrow();
```

**Review by lesshonor — CHANGES_REQUESTED** [review]
I forgot to mention this in my original comment: if you really only want to provide this keymap for `nanoboot`, you can put it into a `keymaps` folder under `nanoboot`—i.e., `keyboards/tweetydabird/lotus58/nanoboot/keymaps/vial`.
But if I was you I would just selectively disable more features in the `rules.mk`. E.g.
```make
# Reduce size on atmega32u4
ifeq ($(strip $(MCU)), atmega32u4)
    TAP_DANCE_ENABLE = no
    QMK_SETTINGS = no
    KEY_OVERRIDE_ENABLE = no
    COMBO_ENABLE = no
    # Reduce size even more for non-nanoboot bootloaders
    ifneq ($(strip $(BOOTLOADER_SIZE)), 512)
        MAGIC_ENABLE = no
        MOUSEKEY_ENABLE = no
    endif
endif
```

**Comment by TweetyDaBird** [comment]
Yep, working on it and testing. 

Unfortunately a trip to the vet with the family cat got in between. Feeling a bit low right now, so I'll pick up later.

File: keyboards/tweetydabird/lotus58/keymaps/vial/vial.json [hunk 10]
```diff
@@ -0,0 +1,420 @@
+{
+	"name": "Lotus 58 Glow VIAL",
+	"author": "Markus Knutsson",
+
```
Comment (lesshonor) [inline keyboards/tweetydabird/lotus58/keymaps/vial/vial.json:4]:
```suggestion
```
Technically not necessary for Vial.

**Review by lesshonor — APPROVED** [review]
The keyboard-level change here is quite reasonable as the same change is being made in [an in-progress QMK PR](https://github.com/qmk/qmk_firmware/pull/22424).

File: keyboards/tweetydabird/lotus58/keymaps/vial/vial.json (hunk 10 above)
Comment (TweetyDaBird) [reply keyboards/tweetydabird/lotus58/keymaps/vial/vial.json:4]: No, not necessary, but very beneficial as it shows as the device name in most OS's. Meaning it's easy to identify if a keyboard is flashed with vial or QMK.

Comment (lesshonor) [reply keyboards/tweetydabird/lotus58/keymaps/vial/vial.json:4]: That isn't determined by `vial.json`, but by `info.json`. [The Vial GUI concatenates `manufacturer` with `keyboard_name`](https://github.com/vial-kb/vial-gui/blob/f5ef91fc30915c6d6946f86d80bd2711e7de4463/src/main/python/vial_device.py#L54).

I flashed an Elite-Pi with this PR.

(I hope your cat is doing better. If that isn't the case, my sympathies.)

Comment (TweetyDaBird) [reply keyboards/tweetydabird/lotus58/keymaps/vial/vial.json:4]: Hmmm.  Ok. Then we are editing that. 

No, unfortunately after 17 years of life, out of which 8 being my daughters life long companion/slash teddy bear/snuggle cat, she past. So yeah, cry fest.

**Pull request merged**